app/helpers/random_assign_helper.rb

```diff
@@ -0,0 +1,174 @@
+require 'set'
+
+class UnableToRandomlyAssignGroupException < Exception
+end
+
+module RandomAssignHelper
+  # Performs the random assignment, and stores them in the database.
+  # - pr_assignment: The peer review assignment.
+  # - num_groups_for_reviewers: How many PR's each group in reviewer_groups should have.
+  def perform_random_assignment(pr_assignment, num_groups_for_reviewers)
+    @eligible_reviewers = pr_assignment.valid_groupings
+    @shuffled_reviewees = []
+    @shuffled_reviewees_pr = []  # A list of peer reviews that match the indices for @shuffled_reviewees
+    @num_groups_for_reviewers = num_groups_for_reviewers
+    @reviewers_assigned_to = Hash.new { |h, k| h[k] = Set.new }  # reviewer_id => set(reviewee_ids)
+
+    reviewer_groups_relation = pr_assignment.valid_groupings
+    reviewee_groups_relation = pr_assignment.parent_assignment.valid_groupings
+
+    @eligible_reviewers.each { |reviewer|  @reviewers_assigned_to[reviewer.id] = Set.new }
+
+    generate_shuffled_reviewees(reviewer_groups_relation, reviewee_groups_relation)
+    remove_groups_from_shuffle_having_peer_review(pr_assignment)
+    perform_assignments()
+  end
+
+  private
+  # Make sure there's enough reviewee groups in the following list such that
+  # we have enough repeats of 'reviewee_groups' that we can assign every
+  # reviewer group at least 'num_groups_for_reviewers' groups.
+  def generate_shuffled_reviewees(reviewer_groups, reviewee_groups)
+    num_times_to_add_reviewee = (reviewer_groups.size.to_f * @num_groups_for_reviewers / reviewer_groups.size).ceil
+    num_times_to_add_reviewee.times { @shuffled_reviewees += reviewee_groups }
+    @shuffled_reviewees.shuffle
+
+    @shuffled_reviewees_pr = @shuffled_reviewees.map { nil }
```

Comment:
Put this in `perform_random_assignments` after you've removed the existing peer reviews. It'll save the call to `pop` below